### PR TITLE
feat: [CDS-45872]: add bypass parameters to Merge PR step

### PR DESCRIPTION
**Pull request by meenaravichandran1**
git cherry-pick bc904fe79e677b98cecc4a345bb67d8d27908a2d

## Description
Add parameters for bypass policy to merge PR step

## Testing Steps
Tested locally and it works as expected.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/39909)

### CODE REVIEW

**Review by rohitk-h — CHANGES_REQUESTED** [review]
> Reviewed 7 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39909)* status: 1 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/gitops/MergePRStepParams.java` line 24 at r1](https://reviewable.io/reviews/harness/harness-core/39909#-NGUCOJoEM8f6SayGbUV:-NGUCOJoEM8f6SayGbUW:b-o1blqc) ([raw file](https://github.com/harness/harness-core/blob/9f42f47b63fe3610c52d31f41f52615cc7f67883/125-cd-nextgen/src/main/java/io/harness/cdng/gitops/MergePRStepParams.java#L24)):*
> ```Java
>   @Builder(builderMethodName = "infoBuilder")
>   public MergePRStepParams(ParameterField<List<TaskSelectorYaml>> delegateSelectors,
>       ParameterField<Boolean> deleteSourceBranch, Map<String, Object> variables) {
> ```

nitpick: We could name this apiParamOptions instead. 
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/gitapi/GitApiTaskParams.java` line 69 at r1](https://reviewable.io/reviews/harness/harness-core/39909#-NGUCTqh9k7PnMlG0Pwo:-NGUCTqh9k7PnMlG0Pwp:b-o1blqc) ([raw file](https://github.com/harness/harness-core/blob/9f42f47b63fe3610c52d31f41f52615cc7f67883/950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/gitapi/GitApiTaskParams.java#L69)):*
> ```Java
>     this.deleteSourceBranch = deleteSourceBranch;
>     this.ref = ref;
>     this.completionOptions = completionOptions;
> ```

nitpick: We could name this apiParamOptions instead. 
___
*[`960-api-services/src/main/java/io/harness/cistatus/service/azurerepo/AzureRepoServiceImpl.java` line 80 at r1](https://reviewable.io/reviews/harness/harness-core/39909#-NGUCh-r9G5K9mtqYM3p:-NGUCh-r9G5K9mtqYM3q:b4xudwr) ([raw file](https://github.com/harness/harness-core/blob/9f42f47b63fe3610c52d31f41f52615cc7f67883/960-api-services/src/main/java/io/harness/cistatus/service/azurerepo/AzureRepoServiceImpl.java#L80)):*
> ```Java
> 
>     if (completionOptions.get(BYPASS_POLICY) != null) {
>       mergeStrategy.put(BYPASS_POLICY,
> ```

We should put the key of the Map as the key here and the value as value looping through each entry. That way it is generic.

Eg: 



_Code snippet:_
```Java
    for(Map.Entry<String, Object> key: completionOptions.entrySet()){
      mergeStrategy.put(String.valueOf(key),
              (((ParameterField) completionOptions.get(key)).getValue()));
    }
```

**Review by meenaravichandran1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39909)* status: 1 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`960-api-services/src/main/java/io/harness/cistatus/service/azurerepo/AzureRepoServiceImpl.java` line 80 at r1](https://reviewable.io/reviews/harness/harness-core/39909#-NGUCh-r9G5K9mtqYM3p:-NGUIY9W5o01DBBeweyC:bgchcp9) ([raw file](https://github.com/harness/harness-core/blob/9f42f47b63fe3610c52d31f41f52615cc7f67883/960-api-services/src/main/java/io/harness/cistatus/service/azurerepo/AzureRepoServiceImpl.java#L80)):*
<details><summary><i>Previously, meenaravichandran1 (Meena Ravichandran) wrote…</i></summary><blockquote>

I did it this way because these two entries in the map are specific to completionOptions. So, I have only fetched them from the map and set to mergeStrategy. 

If we do it in a generic way, later if we add one more entry to the map, we may have to handle it at that point where we have to separate out the entries for different purposes.
</blockquote></details>

Let me know your thoughts about this.

**Comment by rohitk-h** [comment]
*[`960-api-services/src/main/java/io/harness/cistatus/service/azurerepo/AzureRepoServiceImpl.java` line 80 at r1](https://reviewable.io/reviews/harness/harness-core/39909#-NGUCh-r9G5K9mtqYM3p:-NGUKMuv4AkoxQT_pqUA:bslihn9) ([raw file](https://github.com/harness/harness-core/blob/9f42f47b63fe3610c52d31f41f52615cc7f67883/960-api-services/src/main/java/io/harness/cistatus/service/azurerepo/AzureRepoServiceImpl.java#L80)):*
<details><summary><i>Previously, meenaravichandran1 (Meena Ravichandran) wrote…</i></summary><blockquote>

Let me know your thoughts about this.
</blockquote></details>

Though I understand your point, I think the overall solution seems incomplete. This will only work for Azure Completion Options at the moment. Also the reason why we wanted to have a feature flag cause other providers are not supported. 
So, this code can stay but we'll have to add a Feature Flag so UI can show this only if FF is enabled.

**Comment by meenaravichandran1** [comment]
trigger messagecheck

**Review by meenaravichandran1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39909)* status: 8 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/gitops/MergePRStepParams.java` line 24 at r1](https://reviewable.io/reviews/harness/harness-core/39909#-NGUCOJoEM8f6SayGbUV:-NGXvDaC8dEUpPvVfZ8g:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9f42f47b63fe3610c52d31f41f52615cc7f67883/125-cd-nextgen/src/main/java/io/harness/cdng/gitops/MergePRStepParams.java#L24)):*
<details><summary><i>Previously, rohitk-h (Rohit Karelia) wrote…</i></summary><blockquote>

nitpick: We could name this apiParamOptions instead. 
</blockquote></details>

Done.
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/gitapi/GitApiTaskParams.java` line 69 at r1](https://reviewable.io/reviews/harness/harness-core/39909#-NGUCTqh9k7PnMlG0Pwo:-NGXvCyd3qGuU5viImxo:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9f42f47b63fe3610c52d31f41f52615cc7f67883/950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/gitapi/GitApiTaskParams.java#L69)):*
<details><summary><i>Previously, rohitk-h (Rohit Karelia) wrote…</i></summary><blockquote>

nitpick: We could name this apiParamOptions instead. 
</blockquote></details>

Done.
___
*[`960-api-services/src/main/java/io/harness/cistatus/service/azurerepo/AzureRepoServiceImpl.java` line 80 at r1](https://reviewable.io/reviews/harness/harness-core/39909#-NGUCh-r9G5K9mtqYM3p:-NGUMXJHBoYM9IrqPhoo:b-iyqljq) ([raw file](https://github.com/harness/harness-core/blob/9f42f47b63fe3610c52d31f41f52615cc7f67883/960-api-services/src/main/java/io/harness/cistatus/service/azurerepo/AzureRepoServiceImpl.java#L80)):*
<details><summary><i>Previously, rohitk-h (Rohit Karelia) wrote…</i></summary><blockquote>

Though I understand your point, I think the overall solution seems incomplete. This will only work for Azure Completion Options at the moment. Also the reason why we wanted to have a feature flag cause other providers are not supported. 
So, this code can stay but we'll have to add a Feature Flag so UI can show this only if FF is enabled.  
</blockquote></details>

As per our discussion, leaving this part as it is.

File: 125-cd-nextgen/src/main/java/io/harness/cdng/gitops/MergePRStepParams.java [hunk 1]
```diff
@@ -9,16 +9,21 @@
...
 import lombok.Builder;
+import lombok.Getter;
 
 public class MergePRStepParams extends MergePRBaseStepInfo implements GitOpsSpecParameters {
+  @SkipAutoEvaluation @Getter ParameterField<Map<String, Object>> variables;
   @Builder(builderMethodName = "infoBuilder")
-  public MergePRStepParams(
-      ParameterField<List<TaskSelectorYaml>> delegateSelectors, ParameterField<Boolean> deleteSourceBranch) {
+  public MergePRStepParams(ParameterField<List<TaskSelectorYaml>> delegateSelectors,
+      ParameterField<Boolean> deleteSourceBranch, Map<String, Object> variables) {
```
Comment (rohitk-h) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/gitops/MergePRStepParams.java:24]: Nitpick: this could be renamed to api params too.

**Review by rohitk-h — APPROVED** [review]

**Review by meenaravichandran1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39909)* status: 8 file(s) to review. 3. unresolved discussions. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/gitops/MergePRStepParams.java` line 24 at r6](https://reviewable.io/reviews/harness/harness-core/39909#-NGU7tHx9sTK65mqvyYv-r6-24:-NGYGkJCEKoIZR91cw05:bebrt1w) ([raw file](https://github.com/harness/harness-core/blob/235506fc489800c99211b112d64bf95af3b712c5/125-cd-nextgen/src/main/java/io/harness/cdng/gitops/MergePRStepParams.java#L24)):*
<details><summary><i>Previously, rohitk-h (Rohit Karelia) wrote…</i></summary><blockquote>

Nitpick: this could be renamed to api params too.
</blockquote></details>

Will do

**Comment by rohitk-h** [comment]
*[`960-api-services/src/main/java/io/harness/cistatus/service/azurerepo/AzureRepoServiceImpl.java` line 80 at r1](https://reviewable.io/reviews/harness/harness-core/39909#-NGUCh-r9G5K9mtqYM3p:-NGYKPfj9Y3v5JQe-Abf:b-t77xdl) ([raw file](https://github.com/harness/harness-core/blob/9f42f47b63fe3610c52d31f41f52615cc7f67883/960-api-services/src/main/java/io/harness/cistatus/service/azurerepo/AzureRepoServiceImpl.java#L80)):*
<details><summary><i>Previously, meenaravichandran1 (Meena Ravichandran) wrote…</i></summary><blockquote>

As per our discussion, leaving this part as it is.
</blockquote></details>

agree.

**Review by rohitk-h — APPROVED** [review]
> Reviewed 1 of 1 files at r2, 5 of 8 files at r3, 1 of 3 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39909)* status: :shipit: complete! ✓

**Review by rohitk-h — APPROVED** [review]

**Comment by meenaravichandran1** [comment]
trigger codebasehash

**Review by rohitk-h — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39909)* status: :shipit: complete! ✓

**Comment by meenaravichandran1** [comment]
trigger feature-build

**Pull request closed without merging**